tests/core/test_core_containers.cpp

```diff
@@ -156,4 +156,21 @@ TEST_F(core_container_test, static_array) {
     EXPECT_EQ(test_array4.size(), 0);
     EXPECT_EQ(test_array4.max_size(), 0);
     EXPECT_TRUE(test_array4.empty());
+
+    constexpr vecmem::static_array<int, 0> test_array5;
+    constexpr auto test_array5_size = test_array5.size();
```

Comment:
Maybe 

```cpp
static_assert(test_array5_size == 0)
```

to make sure it's actually executed at compile time?

Same below.

Reply:
:confused: I added these `test_array5_size` variables exactly to force the evaluation of these functions at compile time. The only question here is whether we prefer to see errors show up during the running of the unit test like:

```
[ RUN      ] core_container_test.static_array
/home/krasznaa/ATLAS/projects/vecmem/vecmem/tests/core/test_core_containers.cpp:162: Failure
Expected equality of these values:
  test_array5_size
    Which is: 0
  1
[  FAILED  ] core_container_test.static_array (0 ms)
```

, or during the compilation, like:

```
[ 53%] Building CXX object tests/core/CMakeFiles/vecmem_test_core.dir/test_core_containers.cpp.o
/home/krasznaa/ATLAS/projects/vecmem/vecmem/tests/core/test_core_containers.cpp: In member function ‘virtual void core_container_test_static_array_Test::TestBody()’:
/home/krasznaa/ATLAS/projects/vecmem/vecmem/tests/core/test_core_containers.cpp:162:5: error: static assertion failed
     static_assert(test_array5_size == 1);
     ^~~~~~~~~~~~~
tests/core/CMakeFiles/vecmem_test_core.dir/build.make:103: recipe for target 'tests/core/CMakeFiles/vecmem_test_core.dir/test_core_containers.cpp.o' failed
```

I'm not sure that the latter is that much more descriptive. Plus, it prevents us from seeing any other test failures. So all in all, I'm leaning towards keeping the test as-is... :thinking:

Reply:
I'm not sure assigning to a `constexpr` variable actually forces the compiler to execute the expression at compile time, or if it only suggests it. 

If you're certain it does, than let's keep it like it is, if not let's use `static_assert`.

Reply:
I'm 99.9% sure that the compiler would let us know if it's not possible to evaluate the

```c++
constexpr auto test_array5_size = test_array5.size();
```

line at compile time. (Unless there's a bug in the compiler.) I do not believe that `constexpr` is a "suggestion" in that sense. If you declared something as a `constexpr`, the compiler better check that it can in principle comply.

I don't even see a situation when a compiler should meaningfully decide that evaluating a `constexpr` variable would be better at runtime for the application's performance. :confused:

Having said this, my main motivation here is consistency. I'm not 1000% sure of how GCC or Clang work internally with `constexpr`. But I don't want to differentiate our tests on just a hunch.

Reply:
99.9% is good enough for me 😋

Reply:
> I don't even see a situation when a compiler should meaningfully decide that evaluating a constexpr variable would be better at runtime for the application's performance.

As we learned the hard way in ACTS, compile-time resources are every bit as finite as run-time ones. If a compiler gets the impression that evaluating something at compile time will consume unreasonable CPU time or RAM, and deferring the evaluation at runtime is possible, it is a good thing that it has the option of doing so.

This is doubly true considering that some compilers' constexpr evaluators are _much_ slower than run-time expression evaluation, as they often go for an interpreter-like strategy with lots of instrumentation. The reason is that non-reproducibility and undefined behavior have worse consequences at compile-time than they do at runtime.

Reply:
Do you know any examples of this? On a theoretical level this is all fine, but given that GCC will happily run out of memory on some of the ATLAS source files, I just fail to see how they would meaningfully make a decision about this particular thing during compilation. :confused:

Producing some simple piece of code that compiles slowly is not much of a challenge. As I learned already during LHC's Run-1, just having a large (like... really large) number of `std::map` operations in a single function can result in a ridiculous memory/CPU usage during the build. So we had to re-design a piece of code back in the day to do certain things at runtime rather than at compile time, just because the price during compilation was enormous. (And the performance hit during runtime was minimal after the re-write.)

But none of that was ever automatic. The only "automatic thing" that landed on my radar not too long ago are these types of things from GCC:

```
/build/atnight/localbuilds/nightlies/Athena/master/athena/PhysicsAnalysis/JetTagging/JetTagValidation/JetTagDQA/src/BTaggingValidationPlots.cxx: In member function 'virtual void JetTagDQA::BTaggingValidationPlots::initializePlots()':
/build/atnight/localbuilds/nightlies/Athena/master/athena/PhysicsAnalysis/JetTagging/JetTagValidation/JetTagDQA/src/BTaggingValidationPlots.cxx:208:8: note: variable tracking size limit exceeded with -fvar-tracking-assignments, retrying without
   void BTaggingValidationPlots::initializePlots(){
        ^~~~~~~~~~~~~~~~~~~~~~~
```

However an example of GCC actually giving up on `constexpr` would be pretty interesting to see.

Reply:
[Here's a simple example](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxHoADqgKhE4MHt6%2BcQlJAkEh4SxRMVy2mPaOAkIETMQEqT5%2BRXaYDskVVQQ5YZHRegqV1bXpDX3twZ353VwAlLaoXsTI7BxoDL2YqrHEANTBBJu0nhCTmyYA7FYaAIKb15vEmARzDJuSZiYAzOcXpwAiJpd/Fx2mxYTGCh2OZwBN1u90eewOk3enx%2BHGmtE4AFZeH4OFpSKhOG5rNZNgpZvNMMczG8eKQCJpUdMANYgDEaAB0ZhObxOAA42XyTlw3mYzPpOJJsQz8ZxeAoQBo6QzpnBYDBEChUCxYnRouRKGhtbqYsAuFwxTRaARovKIBFpRFglUAJ6cWmGtiCADyDFortxvCwIKM4gDpHwd2aADdMPKw2sml5rdKdiVpbQ8BFiC6PFg3bwCMQ8Cx89MqAZgAoAGp4TAAdy9sUY%2BZkghEYnYUlb8iUamlujFBiMKGJln0mflkGmqFiZRWnAAtF68Y1ms4IK4BvVSIFRnkChlEnOt4esgwOvuJsVSi1hieGiUmnPWtUL11CrY7546j1hm/xoU0xknMCx6IWmCLDwaKYlKYYEhwqi8gAbAuSGSJswDIMgmxmpymwQESljWKQmy4IQJBUjSJEeEa9BbOYIqTAWyrMiAkgAJzsicZhIWYUjsXxZjIXy4ocJKpAlmypA4ni8FygqSoBiq6oQEgho6nR%2BoQOpxooEOppmBoRSWtaxC2vaYaOswxD%2Bu6WqegQPp%2BtKQaGMAoZ4hGT54DGcZ4gmyBJoseKpuiYYZlmOYYJBBZFiW3ConwFbVrWDZNjitL8G2ojiF2WU9io6hhroRT6SORFjhFk6HPis7JHGS4ro%2Ba5%2BBuDDuN%2BgxiruuTvhIGKkPER7JPeYpDWe/4HlwA2rs%2BX5pPUYqzbebSTRMM3zT%2B5qfqte59dNQHkqBRTgTF0EcFi0nSvBiEoWhmzIPpOFmOyHJcPhhFWGOpH4EQ9HUkUmw0Rp0SUVMzFKdMCCYEwWAxDVYXiZJioybwcm2Ap9KQ6QLIcVxPF8RxgnCbyolvLBsmyopWiTOdZgU2jVNYzT0wxmZyRsUAA%3D%3D%3D) where both GCC and clang will refrain from evaluating a constexpr at compile time when told that compilation speed is all that matters via `-O0`.

Beyond this, GCC has a notion of "too complicated to constexpr" via this set of flags...

> **-fconstexpr-depth=n**
>
> Set the maximum nested evaluation depth for C++11 constexpr functions to n. A limit is needed to detect endless recursion during constant expression evaluation. The minimum specified by the standard is 512.
>
> **-fconstexpr-loop-limit=n**
>
> Set the maximum number of iterations for a loop in C++14 constexpr functions to n. A limit is needed to detect infinite loops during constant expression evaluation. The default is 262144 (1<<18).
>
> **-fconstexpr-ops-limit=n**
>
> Set the maximum number of operations during a single constexpr evaluation. Even when number of iterations of a single loop is limited with the above limit, if there are several nested loops and each of them has many iterations but still smaller than the above limit, or if in a body of some loop or even outside of a loop too many expressions need to be evaluated, the resulting constexpr evaluation might take too long. The default is 33554432 (1<<25).

...but I don't know what GCC actually does when the limit is reached. Does it simply error out, or does it have the logic in place to back out of const evaluation in that scenario? I personally don't have time to experiment here, but feel free to do so.

In the interest of the GCC developers' sanity, I personally hope that they somehow handle "unnecessary constexpr" as a special case of the general optimizer logic that is used when the compiler attempts to evaluate expressions at compile time : inlining, CSE, constprop, etc... They did so a long time before `constexpr` was a thing, it just happens that `constexpr` is designed to guarantee that if they attempt this optimization, then given enough time it will eventually succeed.

Reply:
Oh, well, it wasn't so hard to try, so I did it anyway.

Both GCC and Clang stop trying to evaluate complicated constexpr and revert to runtime evaluation after a while.

- [Clang doesn't even try very hard before giving up](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAMxmpK4AMngMmAByPgBGmMR6AA6oCoRODB7evgFByamOAmER0SxxCVy2mPYFDEIETMQEmT5%2BgZXV6XUNBEVRsfF6CvWNzdltQ929JWUSAJS2qF7EyOwcaAxDmKqJxADU4QS7MUzoEAe7iQ2suyb%2BACK7ZrM3AOxWGgCCu9/7VLsQl2I11ubluDy4zxMbxMnx%2BcN2xEwBCWDAuVxYt3eXx%2BUIeVSUryx8J%2BiORxFRx1OgNYAFoITdLEcTgD0bsaY9ZpiYdjvrjuXzPtzziwmOEIJDobCSUiUUzTpz/FjcRx5rROABWXh%2BDhaUioTigyzWXYKRbLTAM/w8UgETQq%2BYAaxA6o0ADozC9/C8ABwun0vLiBIJqjiSLV2vWcXgKEAaG12%2BZwWAwRAoVAsRJ0eLkShoDNZhLALhcII0WgEeIxiAxCMxcINACenGtebYggA8gxaE2dbwsCKjOJe6R8IiHHgAG6YGPDraYZBeCsRg5VCO0PAxIHEBseLDN3gEYh4Fj7%2BZUAzABQANTwmAA7u3Eox9zJBCIxOwpK/5Eo1BHdMGhjGNY1j6BuMaQPMqCJDUM40u2/jRlU841C4DDuJ4LR6KE4R9KUAwVHkaQCKMfiESkxEMFM/TlO0KGdMMTSYdkFR2PRAhdI01H4bREwjMxZG2Ix3EzBCCxLCseiHpgqw8KqGrhsO%2BocKo3oAGw0mpki7MAyDILsxbuv8hpWJYpC7LghAkJaFS7B4%2Bb0Hs5iBLMB4Jo6ICSAAnK6LxmGpZhSF5gVmOpPr6JwYakCeLqkNqurKdGsbxr2iYphASB5pmjk5hAWUFigBhGCWGgVGWFbEFWNbDnWzDbi%2BraMAQnbdhG/ZAUOuqjvRU4zrqc4Lkuw4riGurrpuja7rJB5Hie3AqnwF7XreD5Ptq1r8G%2BojiF%2Bm0/io6jDroFRFcBRpmeNEHinqMHpHBCFIR0zgQK4pHYehIkEaQRE1G95H5Okn28ch44cYxf10aDtTCbh0xfXxTFZIJCNA3M4nmlJiLTfJHCanFEbKapGlabsyCnQZZium6XDGSBZkWfgRBOWYVrmfZ2XxDZrkpVoszzAgmAnAM10hlFMVxvFvCJbYyW2qlHneb5/mBd5IVhd6EUcIh%2BNKVGPP2prZiKQlety7z8xTpV6SeUAA%3D%3D)
- GCC tries harder, but [will eventually give up as well](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAMxmpK4AMngMmAByPgBGmMR6AA6oCoRODB7evgFByamOAmER0SxxCVy2mPYFDEIETMQEmT5%2BgZXV6XUNBEVRsfF6CvWNzdltQ929JWUSAJS2qF7EyOwcaAxDmKqJxADU4QS7MUzoEAe7iQ2suyb%2BACK7ZrM3AOxWGgCCu9/7VLsQl2I11ubluDy4zxMbxMnx%2BcN2xEwBCWDAuVxYt3eXx%2BUIeVSUryx8J%2BiORxFRx1OgNYAFoITdLEcTgD0bsaY9ZpiYdjvrjuXzPtzziwmOEIJDobCSUiUUzTpz/FjcRx5rROABWXh%2BDhaUioTigyzWXYKRbLTAM/w8UgETQq%2BYAaxA6o0ADozC9/C8ABwun0vLiBIJqjiSLV2vWcXgKEAaG12%2BZwWAwRAoVAsRJ0eLkShoDNZhLALhcII0WgEeIxiAxCMxcINACenGtebYggA8gxaE2dbwsCKjOJe6R8IiHHgAG6YGPDraYZBeCsRg5VCO0PAxIHEBseLDN3gEYh4Fj7%2BZUAzABQANTwmAA7u3Eox9zJBCIxOwpK/5Eo1BHdMGhjGNY1j6BuMaQPMqCJDUM40u2/jRlU841C4DDuJ4LR6KE4R9KUAwVHkaQCKMfiESkxEMFM/TlO0KGdMMTSYdkFR2PRAhdI01H4bREwjMxZG2Ix3EzBCCxLCseiHpgqw8KqGrhsO%2BocKo3oAGw0mpki7MAyDILsxbuv8hpWJYpC7LghAkJaFS7B4%2Bb0Hs5iBLMB4Jo6ICSAAnK6LxmGpZhSF5gVmOpPr6JwYakCeLqkNqurKdGsbxr2iYphASB5pmjk5hAWUFigBhGCWGgVGWFbEFWNbDnWzDbi%2BraMAQnbdhG/ZAUOuqjvRU4zrqc4Lkuw4riGurrpuja7rJB5Hie3AqnwF7XreD5Ptq1r8G%2BojiF%2Bm0/io6jDroFRFcBRpmeNEHinqMHpHBCFIR0zgQK4pHYehIkEaQRE1G95H5Okn28ch44cYxf10aDtTCbh0xfXxTFZIJCNA3M4nmlJiLTfJHCanFEbKapGlabsyCnQZZium6XDGSBZkWfgRBOWYVrmfZ2XxDZrkpVoszzAgmAnAM10hlFMVxvFvCJbYyW2qlHneb5/mBd5IVhd6EUcIh%2BNKVGPP2prZiKQlety7z8xTpV6SeUAA%3D%3D).

Anyway, for those who think they know better than the compiler whether something needs to be evaluated at compile time or not, there is always C++20's `consteval`...

Reply:
By the way, in case you're curious, the single top reason why compilers go crazy when asked to compile a single very complicated function is that compilers operate at function granularity, and the resource usage of some very important compilation algorithms like register allocation grows super-linearly with the complexity of their input.

For "sane" functions that do little more than calling into other functions that live into separate cpp files, this is not usually a big problem, because compilers can just guesstimate the complexity of compiling the function using heuristics, and stop inlining calees as soon as the extimated complexity grows too large.

The problem, however, is what happens when the function you're calling cannot be "outlined". What if it's just some random inline declaration in a header without an associated "home" cpp flag? Like, say, most class/function template declarations?

In principle, in that event, I imagine the compiler should be able to do the moral equivalent of "outlining" that function anyway, using some sort of static linkage trick so that the build doesn't later fail due to all the duplicate declarations that will inevitably result across multiple compilation units. But in practice, I don't know of a single compiler that does this.

Maybe it's considered bad style to have multiple clones of the same symbol in the output file even if you can hack the linker into accepting to live with it (I can certainly imagine that it would make the work of tools like debuggers and profilers a lot more "interesting"). Or maybe I'm overestimating the capabilities of modern linkers, and it's just plain verboten to link together multiple instances of the same symbol even if they "hidden" from each other via static linkage.